extensions/src/main/scala/ai/h2o/sparkling/extensions/rest/api/ImportFrameHandler.scala

```diff
@@ -31,10 +35,50 @@ class ImportFrameHandler extends Handler {
   def finalize(version: Int, request: FinalizeFrameV3): FinalizeFrameV3 = {
     val rowsPerChunk = Base64Encoding.decodeToLongArray(request.rows_per_chunk)
     val columnTypes = Base64Encoding.decode(request.column_types)
-    ChunkUtils.finalizeFrame(request.key, rowsPerChunk, columnTypes, null)
+    val frameKey = Key.make(request.key)
+
+    // Collect global domains from local domains on h2o nodes.
+    val collectDomainsTask = new CollectCategoricalDomainsTask(frameKey)
+    collectDomainsTask.doAllNodes()
+    val stringDomains = collectDomainsTask.getDomains()
+    val domains = expandDomains(stringDomains, columnTypes)
+
+    ChunkUtils.finalizeFrame(request.key, rowsPerChunk, columnTypes, domains)
+
+    // Update categorical indexes for all chunks.
+    val frame = DKV.getGet[Frame](frameKey)
+    val categoricalColumnIndices = for ((vec, idx) <- frame.vecs().zipWithIndex if vec.isCategorical) yield idx
```

Comment:
It took my some time to understand what is the purpose of `UpdateCategoricalIndicesTask` but still not rock solid.

Is it because in Sparkling Water we send local indices on each partition and alter we need to re-map to global one?

Reply:
yes, On spark site we calculate local indices for each chunk. `CollectCategoricalDomainsTask` calculates global domain for the whole cluster.  `UpdateCategoricalIndicesTask` translates indexes to the global domain as you said.

Reply:
Thank you!